### PR TITLE
Fix recursive input schema

### LINKED ISSUE #527: Circular class references cause StackOverflowError on InputType generation
**Library Version**
1.4.2

**Describe the bug**
Circular class references are explicitly supported by GraphQL and graphql-kotlin. They appear to work fine when generating the GraphQLType, but fail when generating a GraphQLInputType for the same object. Once a class with a circular reference is introduced to the schema as an input to a Query, the schema generation loops while generating the input type and crashes with a StackOverflowError.

**To Reproduce**
Steps to reproduce the behavior. Please provide:
1. Add a circular reference to a class picked up for schema generation. The simplest way to do this is to add a property of the same type to the class. See example below.
2. Add a query funcition that accepts this class as a parameter
3. Generate schema and notice the output loops and overflows

Minimal repro project available here: https://github.com/cwendel16834/graphqlbugrepro

Repro code
```
class Person {
    val friend: Person = Person()
}

@Component
class Queries : Query {
    fun getPerson(): Person = TODO()

    // Comment this line and it works
    fun addPerson(inputPerson: Person): String = TODO()
}
````

Here is a sample of the output stack trace:
```
at com.expediagroup.graphql.generator.types.InputPropertyBuilder.inputProperty$graphql_kotlin_schema_generator(InputPropertyBuilder.kt:37) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.SchemaGenerator.inputProperty(SchemaGenerator.kt:114) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.types.InputObjectBuilder.inputObjectType$graphql_kotlin_schema_generator(InputObjectBuilder.kt:41) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.SchemaGenerator.inputObjectType(SchemaGenerator.kt:111) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.getGraphQLType(TypeBuilder.kt:84) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.objectFromReflection(TypeBuilder.kt:67) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.graphQLTypeOf$graphql_kotlin_schema_generator(TypeBuilder.kt:46) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.types.InputPropertyBuilder.inputProperty$graphql_kotlin_schema_generator(InputPropertyBuilder.kt:37) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.SchemaGenerator.inputProperty(SchemaGenerator.kt:114) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.types.InputObjectBuilder.inputObjectType$graphql_kotlin_schema_generator(InputObjectBuilder.kt:41) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.SchemaGenerator.inputObjectType(SchemaGenerator.kt:111) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.getGraphQLType(TypeBuilder.kt:84) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.objectFromReflection(TypeBuilder.kt:67) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.TypeBuilder.graphQLTypeOf$graphql_kotlin_schema_generator(TypeBuilder.kt:46) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
	at com.expediagroup.graphql.generator.types.InputPropertyBuilder.inputProperty$graphql_kotlin_schema_generator(InputPropertyBuilder.kt:37) ~[graphql-kotlin-schema-generator-1.4.2.jar:na]
```

**Expected behavior**
Circular references should work on InputTypes the same way they work on normal generated Types. They should not fail to generate.

### PR DESCRIPTION
**Pull request by smyrick**
### :pencil: Description
If there was an input type used recursively or an output type that self-references on one of it's functions then the schema generator would fail. This PR fixes that by doing a couple things:

 * Refactor the schema generator to expose graphQLTypeOf and delete TypeBuilder.
 * Adds unit tests to verify all recursive inputs
 * Caches values and types under construction by considering if they are input types

There are a lot more changes with the schema generator refactor here that I will continue with in a second PR to clean up the code even more, but that will have no effect on features so I am making that a separate PR.

### :link: Related Issues
* Fixes #527
* Helps start resolving https://github.com/ExpediaGroup/graphql-kotlin/issues/84 (more to come on this)

### CODE REVIEW

**Comment by smyrick** [comment]
After this is merged, I can refactor all the type builders (`ArgumentBuilder`, `ObjectBuilder`) to be simple pure functions that accept the `SchemaGenerator` as one of their arguments so they are easier to test and we don't have to mock all the functions in the schema generator to run unit tests. 

This way we can just call them directly from tests and pass in a schema generator in whatever state we want to verify the behaviour. The functions will still be recursive through the `SchemaGenerator` but at least they will be less cyclic with the `TypeBuilder` gone

File: graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/types/InterfaceBuilderTest.kt [hunk 1]
```diff
@@ -67,17 +64,4 @@ internal class InterfaceBuilderTest : TypeTestHelper() {
         assertEquals(1, result?.directives?.size)
         assertEquals("simpleDirective", result?.directives?.first()?.name)
     }
-
-    @Test
-    fun `verify interface is build only once`() {
```
Comment (dariuszkuc) [inline graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/types/InterfaceBuilderTest.kt]: removal of this test (and similar ones for union and objects) implies we no longer cache them?

File: graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/FederatedSchemaGenerator.kt [hunk 2]
```diff
@@ -40,24 +40,15 @@ open class FederatedSchemaGenerator(generatorConfig: FederatedSchemaGeneratorCon
...
             .map { it.loadClass().kotlin }
-            .map {
-                val graphQLType = if (it.isAbstract) {
-                    interfaceType(it)
-                } else {
-                    objectType(it)
-                }
-
-                // workaround to explicitly apply validation
-                config.hooks.didGenerateGraphQLType(it.createType(), graphQLType)
```
Comment (dariuszkuc) [inline graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/FederatedSchemaGenerator.kt]: workaround is no longer needed?

File: graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/state/TypesCache.kt [hunk 3]
```diff
@@ -90,17 +90,23 @@ internal class TypesCache(private val supportedPackages: List<String>) {
...
+            cachedType != null -> cachedType
+            typesUnderConstruction.contains(cacheKey) -> GraphQLTypeReference.typeRef(kClass.getSimpleName(inputType))
             else -> {
-                typeUnderConstruction.add(kClass)
+                typesUnderConstruction.add(cacheKey)
                 val newType = build(kClass)
-                val key = TypesCacheKey(kClass.createType(), false)
-                put(key, KGraphQLType(kClass, newType))
-                typeUnderConstruction.remove(kClass)
+                if (newType !is GraphQLTypeReference) {
```
Comment (dariuszkuc) [inline graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/state/TypesCache.kt]: guess this is the fix for the #527 ?

File: graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/types/InterfaceBuilderTest.kt (hunk 1 above)
Comment (smyrick) [reply graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/types/InterfaceBuilderTest.kt]: We still cache them but now only in the method `buildIfNotUnderConstruction` this simplifies the logic and makes sure we cache only once

File: graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/FederatedSchemaGenerator.kt (hunk 2 above)
Comment (smyrick) [reply graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/FederatedSchemaGenerator.kt]: Correct, as we now call `grahpQLTypeOf` that triggers all the proper hooks

File: graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/state/TypesCache.kt (hunk 3 above)
Comment (smyrick) [reply graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/state/TypesCache.kt]: This one line doesn't solve the issue. It was also causing bugs to appear in the federation code which is why I needed to refactor the schema generator to be able to work without having to directly call `interfaceType` and `objectType`

**Pull request merged**